epicsdbbuilder/const_array.py

```diff
@@ -0,0 +1,92 @@
+from decimal import Decimal
+from .recordbase import quote_string
+from .parameter import Parameter
+
+__all__ = ['ConstArray']
+
+
+class ConstArray:
+    """Constant Link Values. EPICS Base 3.16.1 and above.
+
+    Example: PY Source
+    ------------------
+    `r = records.lsi('r', INP=ConstArray(['Plain String not DBLINK']))`
+
+    Example: Generated DB
+    ---------------------
+    `field(INP, ["Plain String not DBLINK"])`
+    """
+
+    def __init__(self, value):
+        """Constructor.
+
+        Parameters
+        ----------
+        value : iterator
+            Homogeneous non-empty list of values.
+        """
+        self.value = value
+
+    @property
+    def value(self):
+        """A list of values.
+        """
+        return self._value
+
+    @value.setter
+    def value(self, value):
+        self._value = self._sanitize(value, 'ConstArray')
+
+    def _sanitize(self, raw_value, context_name):
+        try:
+            value_list = list(raw_value)
+        except TypeError:
+            value_list = None
+        assert value_list is not None, \
```

Comment:
I definitely don't approve of this one!  What you've done here is to catch the exception, set a special sentinel value, and then test for the sentinel to finally raise the exception you want to.  A more honest way to write this would be:
```python
try:
    value = list(value)
except TypeError:
    raise AssertionError('blah blah')
```

Reply:
I am going to update to update the code as recommended. Yet, let me explain the reason for such a structure. 

Raising an exception inside an exception will print out following report:
```
Traceback (most recent call last):
  File "test.py", line 3, in <module>
    value = list(value)
TypeError: 'int' object is not iterable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test.py", line 5, in <module>
    raise AssertionError('blah blah')
AssertionError: blah blah
```

Unlike this, raising exception outside yields just a simple exception record:
```
Traceback (most recent call last):
  File "test.py", line 7, in <module>
    assert value is not None, AssertionError('blah blah')
AssertionError: blah blah
```

I think the latter case (a single exception) is easier to read quickly especially if it happen to be part of a larger block of text (a build log). At the end, failing one python script does not imply all build scripts in all threads will stop immediately. Yet, I understand that this might not be aligned with the intended use. Thus, I will update the code as recommended.

Reply:
I see what you mean about the nested exceptions, that's not so great.  On the whole I'd just go for `value = list(value)` and expose that slightly unfriendly error directly to the user.

Reply:
Ok. I will do.